source/common/protobuf/utility.h

```diff
@@ -303,6 +312,10 @@ class MessageUtil {
       checkForUnexpectedFields(message, validation_visitor, recurse_into_any);
     }
 
+    // Throw an exception if the config has an invalid Duration field. This is needed
```

Comment:
We need to update the comment below about "recurse the message twice", since this adds one more recursion I think.

Reply:
I was wondering whether I should just create a single visitor (MetaVisitor) that combines multiple visitors and executes them. Then there will be a single recursion, but same time-complexity overall. WDYT?

Reply:
I don't know if it's measurably worse with 2 vs 3 recursions. Separating recursion from the functor seems like a good idea though. More importantly, this top-level validator call should never be called recursively since it'll get quadratic.

Reply:
I don't think that the top-level validator is called multiple times.
Refactoring this to single recursion with multiple validators is IMHO better, and I'll add a TODO for a future PR.